apm-agent-plugins/apm-opentelemetry/apm-opentelemetry-metrics-bridge-parent/apm-opentelemetry-metrics-bridge-v1_14/src/main/java/co/elastic/apm/agent/opentelemetry/metrics/bridge/BridgeFactoryV1_14.java

```diff
@@ -0,0 +1,294 @@
+import io.opentelemetry.api.metrics.DoubleCounter;
+import io.opentelemetry.api.metrics.DoubleCounterBuilder;
+import io.opentelemetry.api.metrics.DoubleGaugeBuilder;
+import io.opentelemetry.api.metrics.DoubleHistogram;
+import io.opentelemetry.api.metrics.DoubleHistogramBuilder;
+import io.opentelemetry.api.metrics.DoubleUpDownCounter;
+import io.opentelemetry.api.metrics.DoubleUpDownCounterBuilder;
+import io.opentelemetry.api.metrics.LongCounter;
+import io.opentelemetry.api.metrics.LongCounterBuilder;
+import io.opentelemetry.api.metrics.LongGaugeBuilder;
+import io.opentelemetry.api.metrics.LongHistogram;
+import io.opentelemetry.api.metrics.LongHistogramBuilder;
+import io.opentelemetry.api.metrics.LongUpDownCounter;
+import io.opentelemetry.api.metrics.LongUpDownCounterBuilder;
+import io.opentelemetry.api.metrics.Meter;
+import io.opentelemetry.api.metrics.MeterBuilder;
+import io.opentelemetry.api.metrics.MeterProvider;
+import io.opentelemetry.api.metrics.ObservableDoubleCounter;
+import io.opentelemetry.api.metrics.ObservableDoubleGauge;
+import io.opentelemetry.api.metrics.ObservableDoubleMeasurement;
+import io.opentelemetry.api.metrics.ObservableDoubleUpDownCounter;
+import io.opentelemetry.api.metrics.ObservableLongCounter;
+import io.opentelemetry.api.metrics.ObservableLongGauge;
+import io.opentelemetry.api.metrics.ObservableLongMeasurement;
+import io.opentelemetry.api.metrics.ObservableLongUpDownCounter;
+
+import javax.annotation.Nullable;
+import java.util.Map;
+
+public class BridgeFactoryV1_14 implements BridgeFactory {
+
+    private static volatile BridgeFactoryV1_14 instance;
+
+    private final WeakMap<AttributeKey<?>, ProxyAttributeKey<?>> convertedAttributeKeys;
+    private final WeakMap<io.opentelemetry.api.common.Attributes, ProxyAttributes> convertedAttributes;
+
+    public BridgeFactoryV1_14() {
+        convertedAttributeKeys = WeakConcurrent.buildMap();
+        convertedAttributes = WeakConcurrent.buildMap();
+    }
+
+    public static BridgeFactoryV1_14 get() {
+        if (instance == null) {
+            throw new IllegalStateException("Bridge has not been activated yet!");
+        }
+        return instance;
+    }
+
+    public static void activate(BridgeFactoryV1_14 instanceToActivate) {
+        if (instance != null) {
+            throw new IllegalStateException("Bridge has already been activated!");
+        }
+        instance = instanceToActivate;
+    }
+
+    public final ProxyAttributes convertAttributes(Attributes attributes) {
+        ProxyAttributes cached = convertedAttributes.get(attributes);
+        if (cached == null) {
+            cached = doConvertAttributes(attributes);
+            if (cached != null) {
+                convertedAttributes.put(attributes, cached);
+            }
+        }
+        return cached;
+    }
+
+    @SuppressWarnings({"rawtypes", "unchecked"})
+    protected ProxyAttributes doConvertAttributes(Attributes attributes) {
```

Comment:
> so I assume they are cached/reused typically. Is this guaranteed?

There is zero guarantee for them to be cached. I assume the typical "just get it working" user won't bother about caching them and recreate them whenever a new measurement occurs.

However if a user uses caching here (e.g. because they produce 1 million measurements per second) it would be *really* bad if we don't cache while the user does.

My thought process here was if the user doesn't do caching our cache won't make things much worse: The user-created attributes are garbage collected at some point, then our entry in the map will be garbagecollected afterwards as well. So we double the garbage, but we won't hog memory due to the cache growing out of bounds. 

Is anything about this assumption incorrect / inaccurate? Where do you see the potential risk with the cache growing?

Reply:
No, I agree.
Maybe the use case of a user that "produce 1 million measurements per second" and NOT caching them will see overhead related to management of large weak-key maps, but probably not a real risk.